iree/turbine/kernel/wave/codegen/handlers.py

```diff
@@ -682,12 +683,108 @@ def handle_abs(source: Value, options: WaveCompileOptions) -> OpResult:
+    Then: e = exp2(t) ≈ exp(-2|x|)
+    So, for x ≥ 0: tanh(x) ≈ (1 - e) / (1 + e)
+
+    To avoid division, compute: r0 = reciprocal(1 + e) [using hardware intrinsic] and then: r = (1 - e) * r0
+    For very small |x| (below a threshold), we use: tanh(x) ≈ x
+    Finally, copy the sign of x
+    """
+
+    element_type = get_type_or_element_type(source.type)
+    if _is_float_type(element_type):
+        # a = x (source)
+        a = source
+        
+        # s = |a|
+        s = math_d.absf(a)
+
+        # Constants:
+        # Compute log2(e)
+        log2e_val = math.log2(math.e)
+
+        # Compute factor = -2 * log2(e)
+        factor_val = -2.0 * log2e_val
+        
+        factor = arith_d.ConstantOp(
+            source.type,
+            DenseElementsAttr.get_splat(
+                source.type, get_constant_attr(factor_val, element_type)
+            ),
+        )
+        # t = factor * s
+        t = arith_d.mulf(s, factor, fastmath=get_fast_math_flags(options))
+
+        # e = exp2(t) using fast hardware intrinsic via math_d
+        e = math_d.exp2(t)
+        
+        one = arith_d.ConstantOp(
+            source.type,
+            DenseElementsAttr.get_splat(
+                source.type, get_constant_attr(1.0, element_type)
+            ),
+        )
+
+        # d = e + 1.0
+        d = arith_d.addf(e, one, fastmath=get_fast_math_flags(options))
+
+        # r = reciprocal(d)
+        r = arith_d.divf(one, d, fastmath=get_fast_math_flags(options))
+        
+        neg_one = arith_d.ConstantOp(
+            source.type,
+            DenseElementsAttr.get_splat(
+                source.type, get_constant_attr(-1.0, element_type)
+            ),
+        )
+        
+        # r = e * (-r) + r  (computes (e-1)/(e+1) without direct division)
+        neg_r = arith_d.mulf(r, neg_one, fastmath=get_fast_math_flags(options))
+        temp = arith_d.mulf(e, neg_r, fastmath=get_fast_math_flags(options))
+        r = arith_d.addf(temp, r, fastmath=get_fast_math_flags(options))
+        
+        # # If s < threshold, then r = a (for small x, tanh(x) ~ x)
+        # thresh_val = 4.997253418e-3
+        # thresh = arith_d.ConstantOp(
+        #     source.type,
+        #     DenseElementsAttr.get_splat(source.type, get_constant_attr(thresh_val, element_type))
+        # )
+        # cmp = arith_d.cmpf(arith_d.CmpFPredicate.OLT, s, thresh)
+        # r = arith_d.select(cmp, a, r)
+        
+        # Copy the sign of a to r: r = copysign(r, a)
+        result = math_d.copysign(r, a)
+    else:
+        raise ValidationError(f"Unhandled operand type for tanh_approx: {element_type}")
+    return result
+
 
 @handle_unary_op(tanh)
 def handle_tanh(source: Value, options: WaveCompileOptions) -> OpResult:
     element_type = get_type_or_element_type(source.type)
     if _is_float_type(element_type):
-        result = math_d.tanh(source)
+        result = math_d.tanh(source, fastmath=get_fast_math_flags(options))
```

Comment:
btw you may need to update some other lit tests that does tanh because now we add a "fast_math" flag